src/test/java/io/supertokens/test/jwt/storage/JWTKeyStorageTest.java

```diff
@@ -0,0 +1,271 @@
+                }
+
+                try {
+                    sqlStorage.setJWTSigningKey_Transaction(con, keyToSet);
+                    return false;
+                } catch (DuplicateKeyIdException e) {
+                    return true;
+                }
+            });
+
+            assert success;
+        } else if (storage.getType() == STORAGE_TYPE.NOSQL_1) {
+            JWTRecipeNoSQLStorage_1 noSQLStorage_1 = (JWTRecipeNoSQLStorage_1) storage;
+            JWTSigningKeyInfo keyToSet = new JWTSymmetricSigningKeyInfo("keyId-1234", 1000, "RSA", "somekeystring");
+
+            try {
+                noSQLStorage_1.setJWTSigningKeyInfoIfNoKeyForAlgorithmExists_Transaction(keyToSet);
+            } catch (Exception e) {
+                fail();
+            }
+
+            /*
+             * When using the same algorithm but different information (key string in this case), setting the key to storage should return
+             * false because an older key for the same algorithm is found in storage
+             */
+            JWTSigningKeyInfo keyWithDifferentKeyString = new JWTSymmetricSigningKeyInfo("keyId-1234", 1000, "RSA", "someDifferentKeyString");
+            boolean successForSameObj = noSQLStorage_1.setJWTSigningKeyInfoIfNoKeyForAlgorithmExists_Transaction(keyWithDifferentKeyString);
+            assert !successForSameObj;
+
+            try {
+                /*
+                 * When setting with the same key id but different algorithm, findOneAndUpdate will try to insert but Mongo
+                 * should throw an error because the _id of the new document already exists in storage
+                 */
+                JWTSigningKeyInfo keyToSet2 = new JWTSymmetricSigningKeyInfo("keyId-1234", 1000, "EC", "somekeystring");
+                noSQLStorage_1.setJWTSigningKeyInfoIfNoKeyForAlgorithmExists_Transaction(keyToSet2);
+            } catch (DuplicateKeyIdException e) {
+                // Do nothing
+            }
+        }
+    }
+
+    /**
+     * For NoSQL only
+     *
+     * Simulate a race condition for parallel threads trying to read from the database (which results in no rows being found initially)
+     * and writing to storage and make sure that at the end of execution both threads used the same key and only one thread was
+     * able to write to storage
+     */
+    @Test
+    public void testThatSettingJWTKeysInParallelWorksAsExpectedForNoSQL() throws Exception {
+        String[] args = {"../"};
+        TestingProcessManager.TestingProcess process = TestingProcessManager.start(args);
+        assertNotNull(process.checkOrWaitForEvent(ProcessState.PROCESS_STATE.STARTED));
+
+        for (int i = 0; i < 10; i++) {
+            String algorithm = "alg" + i;
+            String keyId = "key" + i;
+
+            JWTRecipeStorage storage = StorageLayer.getJWTRecipeStorage(process.getProcess());
+
+            if (storage.getType() != STORAGE_TYPE.NOSQL_1) {
+                return;
+            }
```

Comment:
Add comment as to why this is being done.

Reply:
Done, added to the test description above the method declaration